myia/abstract/utils.py

```diff
@@ -321,11 +327,15 @@ def abstract_clone(self, x: object, *args):
 
 @overload  # noqa: F811
 async def concretize_abstract(self, r: Reference):
-    return Reference(
-        r.engine,
-        r.node,
-        await self(r.context)
-    )
+    # There's a bit of a messup here where we can't build Reference using the
+    # constructor because the node in the reference is sometimes unmanaged.
+    # That's likely because of redirections made by the union code in switch.
```

Comment:
That doesn't sound pleasant for the rest of the code either.

Is there really no way to fix this?

Reply:
The issue is that the constructor of Reference does some extra things such as making sure that the context matches the node's graph, which requires the manager, but in some cases some unmanaged graphs may linger in various caches, and we'll try and fail to copy them through this. There are various solutions:

* Try to eliminate these references since they are probably unnecessary, but I believe that would probably be best addressed by a rewrite of the specializer.
* Simplify the constructor of Reference to not do that and use a helper function to do the context matching when necessary. As it turns out, most of the code already uses `engine.ref(...)` instead of `Reference(engine, ...)`, so I could simply move some code from the constructor of `Reference` to `engine.ref`.